### ISSUE TITLE
DesktopGL emulates DX9 half pixel offset

### ISSUE THREAD

**Issue by roy-t** [MonoGame/MonoGame#6619]
#### What operating system are you using:
- Windows, DesktopGL

There is a rendering difference between the DirectX and OpenGL versions of MonoGame when using the BasicEffect. Because the half pixel offset for OpenGL is not taken into account. While this is taken into account for the SpriteBatch:
https://github.com/MonoGame/MonoGame/blob/faea8a6a89504673e2bb7e435f0da8cc513d8c30/Mono
Game.Framework/Graphics/GraphicsDevice.OpenGL.cs#L272

This means that anything rendered with the BasicEffect shader is slightly blurry on OpenGL. For example see these two screenshots:

OpenGL:
![text_a](https://user-images.githubusercontent.com/623634/50658255-11873f80-0f99-11e9-8789-2a56e4bdbc39.PNG)

DirectX:
![text_c](https://user-images.githubusercontent.com/623634/50658257-14823000-0f99-11e9-9809-9ce415e9e2d8.PNG)

I think we should give the user some way to know that the half pixel offset should be set by themselves or change the OpenGL shader for the BasicEffect to fix this.

What do you think? I can make a PR once I've gathered more input :).

**Comment by mrhelmut** [comment]
We could use preprocessing directives to put OpenGL specific stuff in .fx files with ```#ifdef OPENGL``` and rebuild ```BasicEffect.ogl.mgfxo```?

(ImGUI FTW)

**Comment by Jjagg** [comment]
Hold up, there is already a half pixel offset being applied for OpenGL and not for DX. The current behavior is completely wrong, but there's an explanation for how it got like this, from my understanding:

DX9 is the weird one out. All other graphics APIs do it 'right'. XNA didn't. It applied a half pixel offset for SpriteBatch. So MG technically needs the half-pixel offset on all platforms that do **not** use DX9 if we want to be XNA compatible since XNA uses DX9. The DX backend doesn't have the offset applied I think because it evolved from DX9 and it was forgotten, though it should to be XNA compatible (looking back this is part of why #5144 failed). XNA applies half a pixel offset in SpriteBatch so DX behavior was consistent there which we prove in testing. For OpenGL the half pixel offset is applied to every effect that goes through MojoShader. Somewhere along the line someone must've thought that OpenGL spritebatch rendering wasn't consistent with DX, due to that half pixel offset for every effect, so a reverse half pixel offset to spritebatch was added.

So bottom line is, we need to get rid of the half pixel offset in SpriteBatch for OpenGL. We need to decide whether we want to force the half pixel offset for all our platforms to better be compatible with XNA or if we want to move forward and get rid of the hack and fix this issue once and for all.

@tomspilman Not sure if any consoles use DX9 API, maybe Vita?

For reference:
- SpriteBatch 'reverse' half pixel offset '[fix](https://github.com/MonoGame/MonoGame/blob/f60cfad31d1729af73e13a47eefd5ed63cf307d6/MonoGame.Framework/Graphics/SpriteBatch.cs#L143-L147)' (`NeedsHalfPixelOffset` is set to true for OpenGL)
- pixel offset for GL '[fix](https://github.com/MonoGame/MonoGame/blob/6f34eb393aa0ac005888d74c5c4c6ab5615fdc8c/MonoGame.Framework/Graphics/GraphicsDevice.OpenGL.cs#L906-L952)'
- [lines](https://github.com/espes/mojoshader/blob/master/mojoshader.c#L2229-L2236) where the offset variable is added in MojoShader

**Comment by Jjagg** [comment]
@roy-t If we do move away from XNA behavior and make things consistent with our DX implementation, the right fix would be to remove the lines here and leave `posFixup[2]` and `posFixup[3]` at 0:
https://github.com/MonoGame/MonoGame/blob/6f34eb393aa0ac005888d74c5c4c6ab5615fdc8c/MonoGame.Framework/Graphics/GraphicsDevice.OpenGL.cs#L938-L939

If we go the other way around, we need to apply the half pixel offset in SpriteBatch on all platforms and somehow fix the vertex position in all shaders. That sounds a lot more painful to me, and considering the non-SpriteBatch parts of MG's DX implementation have been inconsistent with XNA for so long (and it's kind of our reference implementation) I'm strongly in favor of dropping the offset altogether and XNA compatibility in this regard with it.

**Comment by roy-t** [comment]
Thanks for the detailed information. It is a lot more complex than I first thought then. I just thought it was forgotten for the basic effect. I'll hold off on changing anything until there is a consensus :). But after that the fix should be really small it seems.

**Comment by roy-t** [comment]
@Jjagg after thinking about it a bit more, I'm a bit confused. So if I target Windows+DirectX do I need to add the half pixel offset to my HLSL shaders, just like in the old XNA/DirectX 9 days? And then if the HLSL shader is auto-translated to GLSL it will add an additional ReverseOffset? Or can I forget about the half pixel offset for DirectX? In that case it seems the behavior is already incompatible with XNA, so it might be OK to break it completely there so we can remove the half pixel entirely?

**Comment by Jjagg** [comment]
@roy-t My post was maybe a bit verbose and hard to parse.

The MG OpenGL implementation is consistent with XNA so it'd need half a pixel offset and the MG DX implementation is not consistent with XNA (except for SpriteBatch) so it does not. ImGui doesn't expect a DX9 renderer, so things look blurry when rendered with a DX9 compatible renderer like MG's GL backend.

> In that case it seems the behavior is already incompatible with XNA, so it might be OK to break it completely there so we can remove the half pixel entirely?

Yes, that's exactly my point as well. MG DX has been incompatible with XNA ever since it's using DX11, so let's make everything work that way.

**Comment by mrhelmut** [comment]
Wait, MojoShader adds a pixel offset to vertices, and then MG adds another (doing twice the job), and then the SpriteBatch removes one? Pfeww, funny how it ended up being like this...

> @roy-t If we do move away from XNA behavior and make things consistent with our DX implementation, the right fix would be to remove the lines here and leave `posFixup[2]` and `posFixup[3]` at 0.

I've tested that and removed the ```NeedsHalfPixelOffset``` from OpenGL and everything looks fine to me. SpriteBatch rendering is 1:1 compared to what it is now.

Consoles don't need it and all have DX11/OpenGL style API (so I would expect them to not be XNA accurate as well).

**Comment by Jjagg** [comment]
> Wait, MojoShader adds a pixel offset to vertices, and then MG adds another (doing twice the job), and then the SpriteBatch removes one? Pfeww, funny how it ended up being like this...

Not exactly. MojoShader adds the parameter for offset, MG sets it for all effects and then undoes it in SpriteBatch. That's technically correct behavior to be compatible with XNA. 

@tomspilman Can you weigh in? I'd like to remove the half pixel offset and be done with this mess. If we remove it altogether we'll get consistent behavior on all platforms, SpriteBatch will render like in XNA and other renders will be offset half a pixel compared to XNA (as it is currently in MG DX).

**Comment by mrhelmut** [comment]
MojoShader only exposes the parameter, alright!

Do we agree that both ```posFixup``` and ```NeedsHalfPixelOffset``` need to be removed to have consistent rendering between MG DX and MG GL?

**Comment by Jjagg** [comment]
> Do we agree that both posFixup and NeedsHalfPixelOffset need to be removed to have consistent rendering between MG DX and MG GL?

Yes, that's a fact. Just waiting for Tom to confirm that breaking XNA compatibility is better than hacking every other platform to enforce the half pixel offset.

**Comment by tomspilman** [comment]
So don't have time to read everything up above here.

IMO keeping XNA compatibility is critical.  We should do that wherever we can.

However we have to move forward and fix XNA's sins too.

Ideally there would be some `UseStandardPixelAddressing` type of global we can use to "opt-in" to breaking XNA compatibility.  We would document it as being something one can disable.  Maybe even including it in all the template projects with a nice descriptive comment.

**Comment by Jjagg** [comment]
I'm gonna implement the fix for this. We'll add a `UseStandardPixelAddressing` static boolean property to `GraphicsDeviceManager` that by default is set to `false`. When set to `false` we'll behave like XNA by applying the half-pixel offset, so the default behavior is compatible with XNA. For new projects we'll recommend setting this flag to `true` with documentation and by setting the flag to `true` in the project templates.

Leaving the flag to `false` will render things as they are currently rendered by MG on OpenGL platforms. Setting it to `true` will render things as they are currently rendered by MG on DX platforms. E.g. for ImGui.NET the flag should be set to `true`.

Note that this will change the behavior for DX projects after upgrading. For `SpriteBatch` nothing will change, if `UseStandardPixelAddressing` is set we'll use a reverse offset in SpriteBatch (as is currently done for OpenGL).

**Comment by RUSshy** [comment]
hacks on hack on hack on hacks, time to ditch xna no ?

**Comment by tomspilman** [comment]
@Jjagg - `UseStandardPixelAddressing` was just a off hand guess at a name.  Maybe put a little more thought into it than i did. :)

**Comment by Jjagg** [comment]
Alternatively we could go with something less technical that expresses the intent, like `Dx9Compatible` or `XnaCompatible`. I like `UseStandardPixelAddressing` better than `HalfPixelOffset` or anything, because it implies that non-XNA compatible is the standard currently.

**Comment by roy-t** [comment]
Awesome, sorry for stirring up the beehive, thanks for picking everything up. Since Jjagg is working on this, I'll try and see if there is another (small) issue that I can dedicate a bit of time to in the coming weeks. Thanks so much!

**Comment by Jjagg** [comment]
> Awesome, sorry for stirring up the beehive

Thanks for reporting, this was long overdue so a poke was needed :)

**Comment by anetegithub** [comment]
Hello there! Because I'm using the `MonoGame.Framework.DesktopGL.Core` version, at the moment there is no possibility to build it under `core` from the development branch, whoever can help with the explanation how can I fix this half pixel now? (Through reflection or directly in the fx effect file)

**Comment by Jjagg** [comment]
Closing this because the OpenGL side is fixed, will open a new issue for the DX side.
